### PR TITLE
CASSANDRA-17808 Optionally avoid hint transfer during decommission

### PR DESCRIPTION
**Pull request by maedhroz**
patch by Caleb Rackliffe; reviewed by Jon Meredith and Stefan Miklosovic for CASSANDRA-17808

### CODE REVIEW

**Comment by maedhroz** [comment]
https://app.circleci.com/pipelines/github/maedhroz/cassandra?branch=CASSANDRA-17808

File: test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java [hunk 1]
```diff
@@ -66,9 +98,10 @@ public void shouldStreamHintsDuringDecommission() throws Exception
             
             // Write data using the second node as the coordinator...
             populate(cluster, "decom_hint_test", 2, 0, 128, ConsistencyLevel.ONE);
-            Long totalHints = countTotalHints(cluster);
+            
             // ...and verify that we've accumulated hints intended for node 4, which is down.
-            assertThat(totalHints).isGreaterThan(0);
+            Awaitility.await().until(() -> countTotalHints(cluster.get(2)) > 0);
```
Comment (maedhroz) [inline test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java]: The metric for total hints is updated async from the actual write...

File: conf/cassandra.yaml [hunk 2]
```diff
@@ -99,6 +99,11 @@ max_hints_file_size: 128MiB
 # Disable the option in order to preserve those hints on the disk.
 auto_hints_cleanup_enabled: false
 
+# Enable/disable transfering hints to a peer during decommission. Even when enabled, this does not guarantee
+# consistenty for logged batches, and it may delay decommission when coupled with a strict hinted_handoff_throttle.
```
Comment (smiklosovic) [inline conf/cassandra.yaml]: typo, "consistency"

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 3]
```diff
@@ -4980,9 +4980,19 @@ private void unbootstrap(Runnable onFinish) throws ExecutionException, Interrupt
...
+        }
+        else
+        {
+            DatabaseDescriptor.setHintedHandoffEnabled(false);
+            HintsService.instance.pauseDispatch();
+            HintsService.instance.deleteAllHints();
+        }
 
         // wait for the transfer runnables to signal the latch.
         logger.debug("waiting for stream acks.");
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/service/StorageService.java:4999]: Maybe move this log statement into the block for transferring above?

File: src/java/org/apache/cassandra/service/StorageService.java [hunk 4]
```diff
@@ -4980,9 +4980,19 @@ private void unbootstrap(Runnable onFinish) throws ExecutionException, Interrupt
...
 
-        Future hintsSuccess = streamHints();
+        if (DatabaseDescriptor.getTransferHintsOnDecommission()) 
+        {
+            setMode(Mode.LEAVING, "streaming hints to other nodes", true);
+            hintsSuccess = streamHints();
+        }
+        else
+        {
+            DatabaseDescriptor.setHintedHandoffEnabled(false);
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/service/StorageService.java:4993]: wdyt about a `setMode(Mode.LEAVING, "deleting hints on decommission");` here?

File: test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java [hunk 5]
```diff
@@ -46,6 +50,36 @@
...
+            // Shutdown node 3 so hints can be written against it.
+            cluster.get(3).shutdown().get();
+
+            cluster.coordinator(1).execute("INSERT INTO " + KEYSPACE + ".decom_no_hints_test (key, value) VALUES (?, ?)", TWO, 0, 0);
+            Awaitility.await().until(() -> countTotalHints(cluster.get(1)) > 0);
+            long hintsAfterShutdown = countTotalHints(cluster.get(1));
+            assertThat(hintsAfterShutdown).isEqualTo(1);
+
+            cluster.get(1).nodetoolResult("decommission", "--force").asserts().success();
+            long hintsDeliveredByDecom = countHintsDelivered(cluster.get(1));
```
Comment (jonmeredith) [inline test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java:79]: wdyt about adding a check for `StorageService.instance.operatingMode == DECOMMISSIONED`?

File: src/java/org/apache/cassandra/hints/HintsDispatchExecutor.java [hunk 6]
```diff
@@ -205,21 +205,27 @@ private final class DispatchHintsTask implements Runnable
         private final UUID hostId;
         private final RateLimiter rateLimiter;
 
-        DispatchHintsTask(HintsStore store, UUID hostId)
+        DispatchHintsTask(HintsStore store, UUID hostId, boolean isTransfer)
```
Comment (jonmeredith) [inline src/java/org/apache/cassandra/hints/HintsDispatchExecutor.java:208]: I don't see where this is used.

Comment (maedhroz) [reply src/java/org/apache/cassandra/hints/HintsDispatchExecutor.java:208]: Good catch. It's now wired up in `TransferHintsTask#transfer()`.

File: src/java/org/apache/cassandra/service/StorageService.java (hunk 3 above)
Comment (maedhroz) [reply src/java/org/apache/cassandra/service/StorageService.java:4999]: We're still doing the main un-bootstrap, so I think it's fine where it is.

File: test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java [hunk 7]
```diff
@@ -46,6 +50,36 @@
...
+    public void shouldAvoidHintTransferOnDecommission() throws Exception
+    {
+        try (Cluster cluster = init(builder().withNodes(3)
+                                             .withConfig(config -> config.set("transfer_hints_on_decommission", false).with(GOSSIP))
+                                             .withoutVNodes()
+                                             .start()))
+        {
+            cluster.schemaChange(withKeyspace("CREATE TABLE %s.decom_no_hints_test (key int PRIMARY KEY, value int)"));
+
+            cluster.coordinator(1).execute("INSERT INTO " + KEYSPACE + ".decom_no_hints_test (key, value) VALUES (?, ?)", ALL, 0, 0);
```
Comment (smiklosovic) [inline test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java]: @maedhroz cant be this wrapped with`withKeyspace` too?

File: test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java [hunk 8]
```diff
@@ -46,6 +50,36 @@
...
+            long hintsBeforeShutdown = countTotalHints(cluster.get(1));
+            assertThat(hintsBeforeShutdown).isEqualTo(0);
+            long hintsDelivered = countHintsDelivered(cluster.get(1));
+            assertThat(hintsDelivered).isEqualTo(0);
+
+            // Shutdown node 3 so hints can be written against it.
+            cluster.get(3).shutdown().get();
+
+            cluster.coordinator(1).execute("INSERT INTO " + KEYSPACE + ".decom_no_hints_test (key, value) VALUES (?, ?)", TWO, 0, 0);
```
Comment (smiklosovic) [inline test/distributed/org/apache/cassandra/distributed/test/HintedHandoffAddRemoveNodesTest.java]: same

**Review by jonmeredith — APPROVED** [review]

**Comment by maedhroz** [comment]
Committed as https://github.com/apache/cassandra/commit/d6aee7e08c658db9d394a6b7e3e27791b4d6854f

**Pull request closed without merging**